In [1]:
import numpy as np

***Подберите скорость обучения (alpha) и количество итераций:***

In [180]:
X = np.array([[1, 1, 1, 1, 1,
               1, 1, 1, 1, 1],
              [1, 1, 2, 1, 3, 
               0, 5, 10, 1, 2]])

y = np.array([45, 55, 50, 59, 65,
              35, 75, 80, 50, 60])

In [181]:
calc_mse = lambda y, y_pred: np.mean((y - y_pred)**2)

In [178]:
def gradient_descent(X, y, W, alpha, eps):
    n = X.shape[1]
    err = np.inf
    err_prev = 0
    iters = 0
    while abs(err - err_prev) > eps:
        y_pred = np.dot(W, X)
        err_prev = err
        err = calc_mse(y, y_pred)
        if err == np.inf:
            break
        for j in range(W.shape[0]):
            W[j] -= alpha * (1/n * 2 * np.sum(X[j] * (y_pred - y)))
        iters += 1
    return W, err, iters

In [189]:
alpha = 0.0001
W = np.array([1, 0.5])
W, alpha, n
eps = 1e-8
alpha = 1
iters = np.inf
iters_prev = 0
err = np.inf
while err == np.inf or iters < iters_prev:
    iters_prev = iters
    W_res, err, iters = gradient_descent(X, y, W.copy(), alpha, eps)
    alpha /= 1.001
alpha, err, iters_prev, iters, W_res

/home/kach/projects/gb-ds/ds_env/lib/python3.7/site-packages/ipykernel_launcher.py:1: RuntimeWarning: overflow encountered in square
  """Entry point for launching an IPython kernel.
/home/kach/projects/gb-ds/ds_env/lib/python3.7/site-packages/numpy/core/_methods.py:75: RuntimeWarning: overflow encountered in reduce
  ret = umr_sum(arr, axis, dtype, out, keepdims)


(0.06376033966243874,
 45.93750006222111,
 173,
 173,
 array([47.23184014,  3.91078662]))

In [190]:
gradient_descent(X, y, W.copy(), alpha, eps)

(array([47.23183548,  3.91078251]), 45.937500061293136, 173)

***Ответ***<br>
Наилучшые Альфа и Итерации это 0.064 и 173) <br>
Понятно, что так ничето не далеате и в этом смысла никакого. ))).

*** *2. В этом коде мы избавляемся от итераций по весам, но тут есть ошибка, исправьте ее:***

In [216]:
W = np.array([1, 0.5])
for i in range(1000):
    y_pred = np.dot(W, X)
    err = calc_mse(y, y_pred)

    '''for ii in range(W.shape[0]):
    W[ii] -= alpha * (1/n * 2 * np.sum(X[ii] * (y_pred - y)))'''
    
    W -= (alpha * (1/n * 2 * np.sum(X * (y_pred - y), axis=1)))

    if i % 100 == 0:
        print(i, W, err)

0 [ 8.02638943 22.17851549] 3173.15
100 [47.18959328  3.92223484] 45.938880779685505
200 [47.23209774  3.91072383] 45.93750000129263
300 [47.23214281  3.91071429] 45.937500000000014
400 [47.23214286  3.91071429] 45.93749999999999
500 [47.23214286  3.91071429] 45.9375
600 [47.23214286  3.91071429] 45.93749999999999
700 [47.23214286  3.91071429] 45.93749999999999
800 [47.23214286  3.91071429] 45.93749999999999
900 [47.23214286  3.91071429] 45.93749999999999


***Задание *3: <br>вместо того, чтобы задавать количество итераций, задайте условие остановки алгоритма - когда ошибка за итерацию
начинает изменяться ниже определенного порога (упрощенный аналог параметра tol в линейной регрессии в sklearn).***

см. задачу 1. 